Combining several interpolation functions using a new util function

ValentinGebhart opened the pull request:
Changes proposed in this PR:
- Adaption of the functions `Impact.local_exceedance_impact`, `Hazard.local_exceedance_intensity`, `Hazard.local_return_period` to use the new util module `climada.util.interpolate` and the three functions therein `interpolate_ev`, `stepfunction_ev` and `group_frequency`.
This makes the functions `local_exceedance_imp`, `loc_return_imp`, `_cen_return_imp`, `local_exceedance_inten`, `_loc_return_inten`, `_cen_return_inten` and `_loc_return_period` obsolete.
- A deprecation warning for the old versions `Impact.local_exceedance_imp`, `Hazard.local_exceedance_inten` is added. 
- All methods `Impact.local_exceedance_impact`, `Hazard.local_exceedance_intensity`, `Hazard.local_return_period` now output a geodataframe, such that one can use the util function `plot_from_gdf` for plotting. This makes the plotting functions `Hazard.plot_rp_intensity` and `Impact.plot_rp_imp` obsolete. A deprecation warning was added.
-  In the methods `Impact.local_exceedance_imp`, `Hazard.local_exceedance_inten`, `Hazard.local_return_period`, before interpolation, we group together values with the same intensity/impact (by summing their frequencies). This was not done in the functions before

Open questions/further possibilities:
- Do we want to try a more efficient output type than gdf if the above function output gdfs with a lot of zeros (e.g. spare gdf etc)?
- Could one do the fitting matrix-wise (or block-wise) instead of column-wise for better efficiency? Not clear due to requirements of underlying fit functions.
- Future Option: PR to add a fit method (to some extreme value distribution, specified by the user).
- Future Option: One could use the util method also in `Impact.calc_freq_curve` to give it a bit more flexibility. Not clear if needed.
- Future Option: We could now easily add a method `Impact.local_return_period`, similar to the Hazard case. Not sure if needed.

This PR fixes #904, #915 and partially also #209

### PR Author Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] Correct target branch selected (if unsure, select `develop`)
- [x] Descriptive pull request title added
- [x] Source branch up-to-date with target branch
- [x] [Documentation](https://climada-python.readthedocs.io/en/latest/guide/Guide_PythonDos-n-Donts.html#2.--Commenting-&-Documenting) updated
- [x] [Tests][testing] updated
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]
- [x] [Changelog](https://github.com/CLIMADA-project/climada_python/blob/develop/CHANGELOG.md) updated

### PR Reviewer Checklist

- [x] Read the [Contribution Guide][contrib]
- [x] [CLIMADA Reviewer Checklist](https://climada-python.readthedocs.io/en/latest/guide/Guide_Review.html#reviewer-checklist) passed
- [x] [Tests][testing] passing
- [x] No new [linter issues][linter]

[contrib]: https://github.com/CLIMADA-project/climada_python/blob/develop/CONTRIBUTING.md
[testing]: https://climada-python.readthedocs.io/en/latest/guide/Guide_Testing.html
[linter]: https://climada-python.readthedocs.io/en/latest/guide/Guide_continuous_integration_GitHub_actions.html#static-code-analysis

chahank commented on climada/engine/impact.py:
```diff
@@ -487,23 +493,51 @@ def local_exceedance_imp(self, return_periods=(25, 50, 100, 250)):
+            )
+
+        # create the output GeoDataFrame
+        gdf = gpd.GeoDataFrame(geometry = [Point(x, y) for y, x in self.coord_exp], crs = self.crs)
```
I think the following is more efficient and clean:

https://geopandas.org/en/stable/docs/reference/api/geopandas.points_from_xy.html

chahank commented on climada/engine/impact.py:
```diff
@@ -813,7 +847,8 @@ def plot_basemap_impact_exposure(self, event_id=1, mask=None, ignore_zero=False,
         return axis
-
+    
+    # TODO: replace with subplots_from_gdf()
```
The name `subplots_from_gdf` is a bit confusing to me. Why would this be a `subplot` instead of a `plot`?

ValentinGebhart replied:
The `subplots_from_gdf` already exists, it is how we now plot local return periods for given hazard intensity thresholds. I named it subplots because it produces a plot with several subplots, each for a different column of the gdf. I am happy to change the name though, just "plot_from_gdf" or "plots_from_gdf"?

chahank replied:
I would then prefer `plot_from_gdf`. Plot is quite clear, and can contain more than one. Plurals are in my opinion to be avoided as much as possible as they easily create confusion.

ValentinGebhart commented:
These are some examples of the behaviour with different settings of `local_exceedance_intensity` and `local_return_period` of a Hazard object, where I plot the interpolation curves for a centroid with few events and with many events (the two examples are from different hazards). 

Note that in contrast to the loglog extrapolation and the stepfunction methods, both the lin-log 1-d fit method from before and the lin-lin extrapolation can produce negative values.

## Few events on the centroid

#### `local_exceedance_intensity` with different settings, and compared to calculation before
![few_events_exceedance_intensity](https://github.com/user-attachments/assets/8db4a0f5-68b6-4796-ba71-98a2f332ea28)

#### `local_return_period` with different settings, and compared to calculation before 
![few_events_return_period](https://github.com/user-attachments/assets/1ff7f0be-1afa-4eb5-86ee-2781cad9da71)

## Many events on the centroid

#### `local_exceedance_frequency` with different settings, and compared to calculation before
![many_events_exceedance_intensity](https://github.com/user-attachments/assets/09b205f8-4b78-4333-a174-65876266e4b1)

#### `local_return_period` with different settings, and compared to calculation before
![many_events_return_period](https://github.com/user-attachments/assets/aa091269-5462-4235-8a8c-21f3fbc83949)

ValentinGebhart commented on climada/engine/impact.py:
```diff
@@ -814,8 +899,16 @@ def plot_basemap_impact_exposure(self, event_id=1, mask=None, ignore_zero=False,
 
         return axis
 
+    # TODO: replace with plot_from_gdf()
```
I would to this in a new PR

ValentinGebhart commented on climada/engine/impact.py:
```diff
@@ -1384,36 +1478,6 @@ def run(i_time):
         return imp_list
 
-#TODO: rewrite and deprecate method
-    def _loc_return_imp(self, return_periods, imp, exc_imp):
```
yes, that's why I didn't add one, also because the only function that used this private method is now not using it anymore.

chahank commented:
This is a great PR and we are looking forward to it being merged. @ValentinGebhart : is there some blocker for which you would need some input?

ValentinGebhart commented:
Main changes since the last review: 
- added option "extrapolate_constant" that interpolates between the points and outside extrapolates by constants (this functionality was called with "interpolate" before)
- method "interpolate" now outputs NaNs for test values outside the seen range
- changed default option to "interpolate"
- added wrapper function `process_and_interpolate_ev` (plus unit test) to util.interpolate.py that combines preproessing and calling the interpolation functions, which before was done in all upper-level functions separately. 
- added util function `convert_frequency_unit_to_time_unit` (plus unit test) to util.checker.py
- moved all existing tests to unit tests
- added one integration test running the functionality for a larger hazard object in test_hazard.py

Note that currently the tests are not passing because some mamba environment cannot be built. If this remains we should soon ask Emanuel for help.

sarah-hlsn commented on climada/util/interpolation.py:
```diff
@@ -24,11 +24,128 @@
+    frequency : array_like
+        1-D array of frequencies to be interpolated.
+    values : array_like
+        1-D array of values (e.g., intensities) to be interpolated.
```
```suggestion
        1-D array of values (e.g., intensities or impacts) to be interpolated.
```

sarah-hlsn commented on climada/util/interpolation.py:
```diff
@@ -24,11 +24,128 @@
+    Parameters
+    ----------
+    test_frequency : array_like
+        1-D array of test frequencies for which values (e.g., intensities) should be assigned.
```
```suggestion
        1-D array of test frequencies for which values (e.g., intensities or impacts) should be assigned.
```

sarah-hlsn commented on climada/util/interpolation.py:
```diff
@@ -24,11 +24,128 @@
+    test_frequency : array_like
+        1-D array of test frequencies for which values (e.g., intensities) should be assigned.
+    test_values : array_like
+        1-D array of test values (e.g., intensities) for which frequencies should be assigned.
```
```suggestion
        1-D array of test values (e.g., intensities or impacts) for which frequencies should be assigned.
```

sarah-hlsn commented on climada/util/interpolation.py:
```diff
@@ -24,11 +24,128 @@
+        If set to True, values (e.g., intensities) are interpolated on log scale.
+        Defaults to False.
+    value_threshold : float, optional
+        Lower threshold to filter values (e.g., intensities). Defaults to None.
```
```suggestion
        Lower threshold to filter values (e.g., intensities or impacts). Defaults to None.
```

sarah-hlsn reviewed (approved):
I would add "impacts" to the examples given for values in the docstrings; since there are just two options for values I think we might as well mention both. Other than that everything looks good to me!

emanuel-schmid commented on climada/hazard/plot.py:
```diff
@@ -32,6 +36,7 @@ class HazardPlot:
     Contains all plotting methods of the Hazard class
     """
 
+    # TODO depreciating warning, to be replaced with plot_from_gdf
```
But strictly speaking this isn't a TODO anymore?

emanuel-schmid commented on climada/hazard/plot.py:
```diff
@@ -41,6 +46,15 @@ def plot_rp_intensity(
         adapt_fontsize=True,
         **kwargs,
     ):
+        """
+        This function is deprecated,
+        use Impact.local_exceedance_impact and util.plot.plot_from_gdf instead.
+        """
+        LOGGER.warning(
+            "The use of Hazard.plot_rp_intensity is deprecated."
+            "Use Hazard.local_exceedance_intensity and util.plot.plot_from_gdf instead."
+        )
```
I'd suggest to use the `@deprecated` decorator instead, like e.g., in the `climada.hazard.centr` module.
I know, atm, the climada codebase is heterogeneous in this regard but in the long run the decorator is meant to take over.

ValentinGebhart replied:
thank you, I didn't know about this. I changed this to decorators now

emanuel-schmid commented:
are the changes in the plots relevant? If they aren't I'd suggest to undo them.
(@ValentinGebhart I already did, ready to push, just let me know.)

emanuel-schmid commented on climada/util/checker.py:
```diff
@@ -228,3 +228,32 @@ def prune_csr_matrix(matrix: sparse.csr_matrix):
     matrix.sum_duplicates()
+
+
+def convert_frequency_unit_to_time_unit(frequency_unit):
+    """Converts common frequency units to corresponding time units. Unknown frequency
+    units are converted to "years".
```
wouldn't the `climada.util.dates_times` package be a better match for this method?

emanuel-schmid commented:
Great Job! Many thanks!
